pkg/config/engine/containerd/systemd.go

```diff
@@ -0,0 +1,43 @@
+/**
+# Copyright 2024 NVIDIA CORPORATION
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+**/
+
+package containerd
+
+import (
+	"os"
+	"path/filepath"
+	"sync"
+)
+
+var (
+	runningSystemd bool
+	detectSystemd  sync.Once
+)
+
+// hasSystemd returns whether systemd is running on this system.
+// This is adapted from the containerd implementation.
+// https://github.com/containerd/containerd/blob/1fb1882c7de7239711c343c65fa266a2f1eba9db/internal/cri/server/runtime_config_linux.go#L63-L68
+func (b *builder) hasSystemd() bool {
```

Comment:
Is it possible that a node has a running systemd , but cgroup driver configured for the kubelet is `cgroupfs` ?

Reply:
I would say that it's possible but unlikely with newer containerd / kubelet versions.

As of k8s 1.22 the default value was `systemd` (see https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver/#configuring-the-kubelet-cgroup-driver).

It is also recommended to use the `systemd` driver with `cgroupv2` when using containerd at least: https://kubernetes.io/docs/setup/production-environment/container-runtimes/#containerd-systemd